### ISSUE TITLE
Using a module as a template for multiple applications

### ISSUE THREAD

**Issue by zette** [hashicorp/terraform#6303]
Hello,
I'm using terraform as infrastructure builder on AWS. We are using Elasticbeanstalk applications, that are parametrized through AWS environment variables.

I've been able to implement building EBS instance as resource "aws_elastic_beanstalk_environment" in module included in VPC infrastructure.
I can copy this module, change its name, import it and manually add entries for each application separately with specific entries:
setting {
    namespace = "aws:elasticbeanstalk:application:environment"
    name = "xyz"
    value = "zyx"
  }
... and it works.

But I'd like to use this module as template for all environments adding needed/specific settings to resource when importing the same module files with different names from main package. The parametrization for resource settings passed to module ie.: with variables.

But I can not see any way to pass settings to resource (list, json, import from file...) from outside of it's definition.

Am I trying to do it wrong way? Does anyone have any idea how I can achieve my goal (resource module parametrization) without the need of creating new module for each application?

Regards
Zbyszek

**Comment by dharrisio** [comment]
@zette Thanks for creating an issue for this. This is something that I have come across as well, and unfortunately, you are correct that each module would need to be copied per application. 

There is no way that I know of to pass in a new `setting` from the Terraform document that is including the module.  So the issue here is that the module would need to contain every possible setting that may be used by all of your applications. I made an attempt to create a module like that but I was unable to get it to work. In this case I found that there were really two types of settings. 

- Settings where default values could be set that would be the same as if the setting was omitted. 
- Settings that have no default value - setting these to anything alters how the environment would be configured. A good example of this is the platform specific settings, it would not make sense for java options and a nodejs version to be set in the same environment.

In the second case, I attempted to set a default value of `""`. Unfortunately this doesn't work, because it still gets included in the set of options to be used. The reason this happens is at https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_elastic_beanstalk_environment.go#L468

It checks to see if the value is not nil, which in the case of `""` is true (so it gets included in the set of options). It may be as simple as changing that line to check if the value is not nil or "". 

@catsby does this sound like an appropriate way to deal with settings in a module? Let me know if the goal here is a bit confusing. I should be able to create an example module/terraform document that shows what we're both trying to accomplish here.
